pkg/runtime/runtime.go

```diff
@@ -0,0 +1,269 @@
+	apstore "github.com/argoproj-labs/argocd-autopilot/pkg/store"
+	"github.com/ghodss/yaml"
+	billyUtils "github.com/go-git/go-billy/v5/util"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kusttypes "sigs.k8s.io/kustomize/api/types"
+)
+
+type (
+	Runtime struct {
+		metav1.TypeMeta   `json:",inline"`
+		metav1.ObjectMeta `json:"metadata"`
+
+		Spec RuntimeSpec `json:"spec"`
+	}
+
+	RuntimeSpec struct {
+		DefVersion         *semver.Version `json:"defVersion"`
+		Version            *semver.Version `json:"version"`
+		BootstrapSpecifier string          `json:"bootstrapSpecifier"`
+		Components         []AppDef        `json:"components"`
+	}
+
+	AppDef struct {
+		Name string `json:"name"`
+		Type string `json:"type"`
+		URL  string `json:"url"`
+		Wait bool   `json:"wait"`
+	}
+)
+
+func Download(version *semver.Version, name string) (*Runtime, error) {
+	var (
+		body []byte
+		err  error
+	)
+
+	if strings.HasPrefix(store.RuntimeDefURL, "http") {
+		urlObj, err := url.Parse(store.RuntimeDefURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed parsing url: %w", err)
+		}
+
+		if urlObj.Query().Get("ref") == "" {
+			urlObj.Query().Set("ref", version.String())
+		}
+
+		res, err := http.Get(urlObj.String())
+		if err != nil {
+			return nil, fmt.Errorf("failed to download runtime definition: %w", err)
+		}
+
+		defer res.Body.Close()
+		body, err = ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read runtime definition data: %w", err)
+		}
+	} else {
+		body, err = ioutil.ReadFile(store.RuntimeDefURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read runtime definition data: %w", err)
+		}
+	}
+
+	runtime := &Runtime{}
+	err = yaml.Unmarshal(body, runtime)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal runtime definition data: %w", err)
+	}
+
+	runtime.Name = name
+	runtime.Namespace = name
+	return runtime, nil
+}
+
+func Load(fs fs.FS, filename string) (*Runtime, error) {
+	cm := &v1.ConfigMap{}
+	if err := fs.ReadYamls(filename, cm); err != nil {
+		return nil, err
+	}
+
+	data := cm.Data["runtime"]
+	runtime := &Runtime{}
+	return runtime, yaml.Unmarshal([]byte(data), runtime)
+}
+
+func (r *Runtime) Save(fs fs.FS, filename string) error {
+	data, err := yaml.Marshal(r)
+	if err != nil {
+		return err
+	}
+
+	cm := v1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "codefresh-cm",
+			Namespace: r.Namespace,
+			Labels: map[string]string{
+				apstore.Default.LabelKeyAppManagedBy: store.Get().BinaryName,
+				store.Get().LabelKeyCFType:           "runtimeDef",
+			},
+		},
+		Data: map[string]string{
+			"runtime": string(data),
+		},
+	}
+
+	return fs.WriteYamls(filename, cm)
+}
+
+func (r *Runtime) Upgrade(fs fs.FS, newRt *Runtime) ([]AppDef, error) {
+	newComponents, err := r.Spec.upgrade(fs, &newRt.Spec)
```

Comment:
Creating a new app is a bit more complicated than just deleting it. So I preferred leaving it to autopilot.
In the autopilot, we always do clone->change->commit->push, so I had to call all the `RunAppCreate` after the commit/push of the rest of the changes (update and delete of applications)
I think we should refactor Autopilot to be able to get an existing filesystem, with an already existing repo cloned to it, and then just make the changes, without commit/push at the end. I already made a plan on how to implement it. It will make a lot of our cli-v2 code much simpler.